File: java/590. N-ary Tree Postorder Traversal.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<Integer> postorder(Node root) {
        List<Integer> result = new ArrayList<>();
        Deque<Node> stack = new ArrayDeque<>();
        while (root != null || !stack.isEmpty()) {
            if (root == null) root = stack.pollLast();
            result.add(root.val);
            if (root.children == null) {
                root = null;
                continue;
            }
            for (Node child : root.children) {
                stack.addLast(child);
            }
            root = null;
        }
        for (int i = 0, j = result.size()-1; i < j; i++, j--) {
            int temp = result.get(i);
            result.set(i, result.get(j));
            result.set(j, temp);
        }
        return result;
    }
}

File: go/590. N-ary Tree Postorder Traversal.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Children []*Node
 * }
 */

func postorder(root *Node) []int {
    stack := make([]*Node, 0, 1000)
    result := []int{}
    for root != nil || len(stack) > 0 {
        if root == nil {
            root = stack[len(stack)-1]
            stack = stack[:len(stack)-1]
        }
        result = append(result, root.Val)
        if root.Children == nil {
            root = nil
            continue
        }
        for _, child := range root.Children {
            stack = append(stack, child)
        }
        root = nil
    }
    for i, j := 0, len(result)-1; i < j; i, j = i+1, j-1 {
        result[i], result[j] = result[j], result[i]
    }
    return result
}

